src/contracts/libraries/GPv2Trade.sol

```diff
@@ -0,0 +1,148 @@
+// SPDX-License-Identifier: LGPL-3.0-or-later
+pragma solidity ^0.7.6;
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "./GPv2Order.sol";
+import "./GPv2Signing.sol";
+
+/// @title Gnosis Protocol v2 Trade Library.
+/// @author Gnosis Developers
+library GPv2Trade {
+    using GPv2Order for GPv2Order.Data;
+    using GPv2Order for bytes;
+    using GPv2Signing for GPv2Order.Data;
+
+    /// @dev A struct representing a trade to be executed as part a batch
+    /// settlement.
+    struct Data {
+        uint256 sellTokenIndex;
+        uint256 buyTokenIndex;
+        address receiver;
+        uint256 sellAmount;
+        uint256 buyAmount;
+        uint32 validTo;
+        bytes32 appData;
+        uint256 feeAmount;
+        uint256 flags;
+        uint256 executedAmount;
+        uint256 feeDiscount;
+        bytes signature;
+    }
+
+    /// @dev Recovered trade data containing the extracted order and the
+    /// recovered owner address.
+    struct Recovered {
+        GPv2Order.Data order;
+        bytes orderUid;
+        address owner;
+        uint256 executedAmount;
+        uint256 feeDiscount;
+    }
+
+    /// @dev Returns a new recovered trade with a pre-allocated buffer for
+    /// packing the order unique identifier.
+    function newRecovered() internal pure returns (Recovered memory trade) {
+        trade.orderUid = new bytes(GPv2Order.UID_LENGTH);
+    }
+
+    /// @dev Recovers trade data from the specified settlement trade input.
+    ///
+    /// @param trade Memory location used for writing the recovered trade data.
+    /// @param domainSeparator The domain separator used for signing the order.
+    /// @param tokens The list of tokens included in the settlement. The token
+    /// indices in the trade parameters map to tokens in this array.
+    /// @param input The input settlement trade data to recover the order and
+    /// signer from.
+    function recoverTrade(
+        Recovered memory trade,
+        bytes32 domainSeparator,
+        IERC20[] calldata tokens,
+        GPv2Trade.Data calldata input
+    ) internal view {
+        GPv2Order.Data memory order = trade.order;
+
+        GPv2Signing.Scheme signingScheme =
+            GPv2Trade.extractOrder(input, tokens, order);
```

Comment:
We allocate the order once, then modify it with the side effects of `extractOrder` and as a bonus get `signingScheme`. Can `order` instead be added to the output, since we don't need to change the content of `order` once allocated?

Reply:
Returning memory always allocates.

Reply:
Yes, but allocating is what we do in line 62, so AFAICT it should be fine if the allocation is done directly in `extractOrder`.

Reply:
Since Solidity v0.7.6, `GPv2Order.Data memory order = trade.order;` no longer allocates, it just sets `order` to point to the same memory location as `trade.order`.

We do this because `trade.order` is a `dup, add, mload, mask` on every access instead of just a `dup` - and we use it more than once.

Reply:
I see, I wrongly interpreted `GPv2Order.Data memory order = trade.order;` as copying the data from calldata to memory, but this is indeed not the case since `trade` is our preallocated memory. :+1: